File: src/main/java/org/distril/beengine/inventory/transaction/action/TakeItemStackAction.java
package org.distril.beengine.inventory.transaction.action;

import com.nukkitx.protocol.bedrock.data.inventory.StackRequestSlotInfoData;
import org.distril.beengine.inventory.transaction.ItemStackTransaction;
import org.distril.beengine.material.Material;
import org.distril.beengine.player.Player;

public class TakeItemStackAction extends MoveItemStackAction {

	public TakeItemStackAction(StackRequestSlotInfoData from, StackRequestSlotInfoData to,
	                           ItemStackTransaction transaction, int count, int requestId) {
		super(from, to, transaction, count, requestId);
	}

	@Override
	public boolean execute(Player player) {
		var fromItem = this.getFromItem();
		var toItem = this.getToItem();

		// double-click
		if (toItem.getMaterial() != Material.AIR) {
			// todo check count from and to items
			toItem.setCount(toItem.getCount() + Math.min(fromItem.getCount(), this.getCount()));

			fromItem = null;
		}

		this.setFromItem(toItem);
		this.setToItem(fromItem);
		return true;
	}
}

File: src/main/java/org/distril/beengine/inventory/transaction/action/TakeItemStackAction.kt
package org.distril.beengine.inventory.transaction.action

import com.nukkitx.protocol.bedrock.data.inventory.StackRequestSlotInfoData
import org.distril.beengine.inventory.transaction.ItemStackTransaction
import org.distril.beengine.material.Material
import org.distril.beengine.player.Player
import kotlin.math.min

class TakeItemStackAction(
	from: StackRequestSlotInfoData,
	to: StackRequestSlotInfoData,
	transaction: ItemStackTransaction,
	count: Int,
	requestId: Int
) : MoveItemStackAction(from, to, transaction, count, requestId) {

	override fun execute(player: Player): Boolean {
		var fromItem = this.fromItem
		val toItem = this.toItem!!

		// double-click
		if (toItem.material != Material.AIR) {
			// todo check count from and to items
			toItem.count = toItem.count + min(fromItem!!.count, this.count)
			fromItem = null
		}

		this.fromItem = toItem
		this.toItem = fromItem
		return true
	}
}
